table/queries/derived_satellite_scans.sql

```diff
@@ -0,0 +1,205 @@
+    WHEN ENDS_WITH(error, "read: protocol error") THEN "❗️read/protocol_error"
+    WHEN ENDS_WITH(error, "socket: too many open files") THEN "❗️setup/system_failure"
+    ELSE CONCAT("❗️dns/unknown_error:", error)
+  END
+);
+
+CREATE TEMP FUNCTION InvalidIpType(ip STRING) AS (
+  CASE
+    WHEN STARTS_WITH(ip, "0.") THEN "❗️ip_invalid:zero"
+    WHEN STARTS_WITH(ip, "127.") THEN "❗️ip_invalid:local_host"
+    WHEN STARTS_WITH(ip, "10.") THEN "❗️ip_invalid:local_net"
+    WHEN NET.IP_TO_STRING(NET.IP_TRUNC(NET.SAFE_IP_FROM_STRING(ip), 12)) = "172.16.0.0"  THEN "❗️ip_invalid:local_net"
+    WHEN STARTS_WITH(ip, "192.168.") THEN "❗️ip_invalid:local_net"
+    ELSE NULL
+  END
+);
+
+CREATE TEMP FUNCTION ClassifySatelliteErrorNegRCode(error STRING) AS (
+  CASE
+    WHEN (error IS NULL OR error = "" OR error = "null")  THEN "❗️read/udp.timeout"
+    ELSE ClassifySatelliteError(error)
+  END
+);
+
+CREATE TEMP FUNCTION TlsCertMatchOutcome(domain STRING,
+                                         cert BYTES,
+                                         https_is_known_blockpage BOOLEAN,
+                                         https_page_signature STRING) AS (
+  CASE
+    WHEN (STRPOS(cert, CAST(domain AS BYTES)) > 0
+          OR STRPOS(cert, 
+             CAST(CONCAT("*.", NET.REG_DOMAIN(domain)) AS BYTES)) > 0)
+      THEN "expected/cert_match"
+    WHEN https_is_known_blockpage
+      THEN CONCAT("https_blockpage:", https_page_signature)                                      
+    ELSE "cert_mismatch"
+  END
+);
+
+CREATE TEMP FUNCTION IsCertForDomain(tls_cert BYTES, domain STRING) AS (
+  STRPOS(tls_cert, CAST(domain AS BYTES)) > 0 OR 
+  STRPOS(tls_cert, CAST(CONCAT("*.", NET.REG_DOMAIN(domain)) AS BYTES)) > 0
+);
+
+CREATE TEMP FUNCTION AnswersSignature(answers ANY TYPE) AS (
+  ARRAY_TO_STRING(ARRAY(
+    SELECT DISTINCT
+      CASE
+        WHEN answer.as_name != "" THEN answer.as_name
+        WHEN answer.asn IS NOT NULL THEN CONCAT("AS", answer.asn)
+        ELSE "missing_as_info"
+      END
+    FROM UNNEST(answers) answer
+  ), ",")
+);
+
+
+CREATE TEMP FUNCTION OutcomeString(domain_name STRING,
+                                   dns_error STRING,
+                                   rcode INTEGER,
+                                   answers ANY TYPE) AS (
+    CASE 
+        WHEN dns_error IS NOT NULL THEN ClassifySatelliteError(dns_error)
+        WHEN rcode = -1 THEN ClassifySatelliteErrorNegRCode(dns_error)
+        WHEN rcode != 0 THEN ClassifySatelliteRCode(rcode)
+        WHEN ARRAY_LENGTH(answers) = 0 THEN "❗️answer:no_answer"
+        ELSE IFNULL(
+            (SELECT InvalidIpType(answer.ip) FROM UNNEST(answers) answer LIMIT 1),
+            CASE
+                WHEN (SELECT LOGICAL_OR(answer.matches_control.ip)
+                      FROM UNNEST(answers) answer) THEN "✅answer:matches_ip"
+                WHEN (SELECT LOGICAL_OR(answer.http_analysis_is_known_blockpage)
```

Comment:
Move content-based checks to after the TLS checks

Reply:
The point here is if we see both a known blockpage and tls cert mismatch (or match even) we want to display the blockpage outcome, not the tls outcome.

`❗️page:http_blockpage:c_isp_ru_ertelecom_redirect_301_1_satellite` 
is a more informative outcome to a user about what's happening than
`❗️page:cert_not_for_domain:ERTELECOM`

Reply:
I wouldn't mind having two or three versions of the outcome string though, so we can compare. But I'd like the main one to be the more robust one that respects the abstraction/event order.

Reply:
discussed offline